File: xflaim/src/cs/Backup.cpp
//------------------------------------------------------------------------------
// Desc: Native C routines to support C# Backup class
// Tabs:	3
//
// Copyright (c) 2006-2007 Novell, Inc. All Rights Reserved.
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; version 2.1
// of the License.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Library Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, contact Novell, Inc.
//
// To contact Novell about this file by physical or electronic mail, 
// you may find current contact information at www.novell.com.
//
// $Id$
//------------------------------------------------------------------------------

#include "xflaim.h"

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC void XFLAPI xflaim_Backup_Release(
	IF_Backup *	pBackup)
{
	if (pBackup)
	{
		pBackup->Release();
	}
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC FLMUINT64 XFLAPI xflaim_Backup_getBackupTransId(
	IF_Backup *	pBackup)
{
	return( pBackup->getBackupTransId());
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC FLMUINT64 XFLAPI xflaim_Backup_getLastBackupTransId(
	IF_Backup *	pBackup)
{
	return( pBackup->getLastBackupTransId());
}

typedef RCODE (XFLAPI * BACKUP_CLIENT)(
	const void *	pvData,
	FLMUINT32		ui32DataLen);

/****************************************************************************
Desc:
****************************************************************************/
class CS_BackupClient : public IF_BackupClient
{
public:

	CS_BackupClient(
		BACKUP_CLIENT	fnBackupClient)
	{
		m_fnBackupClient = fnBackupClient;
	}

	virtual ~CS_BackupClient()
	{
	}

	RCODE XFLAPI WriteData(
		const void *			pvBuffer,
		FLMUINT					uiBytesToWrite)
	{
		return( m_fnBackupClient( pvBuffer, (FLMUINT32)uiBytesToWrite));
	}

private:

	BACKUP_CLIENT	m_fnBackupClient;
};

typedef RCODE (XFLAPI * BACKUP_STATUS)(
	FLMUINT64	ui64BytesToDo,
	FLMUINT64	ui64BytesDone);

/****************************************************************************
Desc:
****************************************************************************/
class CS_BackupStatus : public IF_BackupStatus
{
public:

	CS_BackupStatus(
		BACKUP_STATUS	fnBackupStatus)
	{
		m_fnBackupStatus = fnBackupStatus;
	}

	virtual ~CS_BackupStatus()
	{
	}

	RCODE XFLAPI backupStatus(
		FLMUINT64	ui64BytesToDo,
		FLMUINT64	ui64BytesDone)
	{
		return( m_fnBackupStatus( ui64BytesToDo, ui64BytesDone));
	}

private:

	BACKUP_STATUS	m_fnBackupStatus;
};

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_Backup_backup(
	IF_Backup *		pBackup,
	const char *	pszBackupPath,
	const char *	pszPassword,
	FLMUINT32 *		pui32SeqNum,
	BACKUP_CLIENT	fnBackupClient,
	BACKUP_STATUS	fnBackupStatus)
{
	RCODE					rc = NE_XFLM_OK;
	IF_BackupClient *	pBackupClient = NULL;
	IF_BackupStatus *	pBackupStatus = NULL;
	FLMUINT				uiSeqNum;

	if (fnBackupClient)
	{
		if ((pBackupClient = f_new CS_BackupClient( fnBackupClient)) == NULL)
		{
			rc = RC_SET( NE_XFLM_MEM);
			goto Exit;
		}
	}
	if (fnBackupStatus)
	{
		if ((pBackupStatus = f_new CS_BackupStatus( fnBackupStatus)) == NULL)
		{
			rc = RC_SET( NE_XFLM_MEM);
			goto Exit;
		}
	}
	if (RC_BAD( rc = pBackup->backup( pszBackupPath, pszPassword, pBackupClient,
								pBackupStatus, &uiSeqNum)))
	{
		goto Exit;
	}
	if (pui32SeqNum)
	{
		*pui32SeqNum = (FLMUINT32)uiSeqNum;
	}
	
Exit:

	if (pBackupClient)
	{
		pBackupClient->Release();
	}
	
	if (pBackupStatus)
	{
		pBackupStatus->Release();
	}
	
	return( rc);
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_Backup_endBackup(
	IF_Backup *	pBackup)
{
	return( pBackup->endBackup());
}

File: xflaim/src/java/wrapper/xflaim/Backup.java
//------------------------------------------------------------------------------
// Desc:	Backup
// Tabs:	3
//
// Copyright (c) 2003-2007 Novell, Inc. All Rights Reserved.
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; version 2.1
// of the License.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Library Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, contact Novell, Inc.
//
// To contact Novell about this file by physical or electronic mail, 
// you may find current contact information at www.novell.com.
//
// $Id$
//------------------------------------------------------------------------------

package xflaim;
import java.io.FileNotFoundException;

/**
 * This classes provides methods to back up an XFlaim database
 */
public class Backup
{
	// This constructor doesn't need to do much of anything; it's here mostly
	// to ensure that Backup does NOT have a public constructor.  (The
	// application is not supposed to call new on Backup; Backup objects
	// are created by a call to Db.backupBegin
	
	Backup(
		long		lThis,
		Db			jdb)
	{
		m_this = lThis;
		m_jdb = jdb;
	}

	/**
	 * Finalize method used to release native resources on garbage collection.
	 */	
	public void finalize()
	{
		if (m_this != 0)
		{
			_release( m_this);
			m_this = 0;
		}
		
		m_jdb = null;
	}
	
	/**
	 * Get the transaction ID for this backup operation. 
	 * @return Returns the transaction ID for this backup operation.
	 */
	public long getBackupTransId()
	{
		return _getBackupTransId( m_this);
	}
	
	/**
	 * Gets the transaction ID for the last backup job run on this database.
	 * @return Returns the transaction ID for the last backup job run on the
	 * database associated with this Backup object.
	 */
	public long getLastBackupTransId()
	{
		return _getLastBackupTransId( m_this);
	}
	
	/**
	 * Performs the backup operation.  <code>sBackupPath</code> and <code>
	 * backupClient</code> are mutually exclusive.  If <code>backupClient</code> is null,
	 * then an instance of <code>DefaultBackupClient</code> will be created
	 * and <code>sBackupPath</code> passed into its constructor.  If <code>
	 * Client</code> is non-null, <code>sBackupPath</code> is ignored.
	 * @param sBackupPath Optional.  The full pathname of a file to store the
	 * backed up data.
	 * @param backupClient Optional.  If non-null, then it will be used as the backup
	 * client.
	 * @param backupStatus Optional.  If non-null, then <code>backupStatus.backupStatus
	 * </code> will be called periodically to inform the application about the
	 * progress of the backup operation.
	 * @return Returns the sequence number of this backup.  (This is for
	 * informational purposes only; for instance, users can use it to label
	 * their backup tapes.)
	 * @throws XFlaimException
	 */
	public long backup(
		String				sBackupPath,
		String				sPassword,
		BackupClient		backupClient,
		BackupStatus		backupStatus) throws XFlaimException
	{
		BackupClient	backClient;
		
		if (backupClient == null)
		{
			try
			{
				backClient = new DefaultBackupClient( sBackupPath);
			}
			catch (FileNotFoundException e)
			{
				throw new XFlaimException( xflaim.RCODE.NE_FLM_OPENING_FILE,
					"IOException opening " + sBackupPath + ". Message from JVM was" +
					e.getMessage());
			}
		}
		else
		{
			backClient = backupClient;
		}
		
		return _backup( m_this, sBackupPath, sPassword, backClient, backupStatus);
	}

	/**
	 * Ends the backup operation.
	 * @throws XFlaimException
	 */
	public void endBackup() throws XFlaimException
	{
		_endBackup( m_this);
	}
	
	// Reassigns the object to "point" to a new F_Backup instance and a new
	// Db.  Called by any of the member functions that take a
	// Db.backupBegin parameter.  Shouldn't be called by outsiders, so it's
	// not public, but it must be callable for other instances of this class.
	// NOTE:  This function does not result in a call to F_Backup::Release()
	// because that is done by the native code when the F_Backup object is 
	// reused.  Calling setRef() in any case except from within
	// Db.backupBegin will result in a memory leak on the native side!
	void setRef(
		long		lBackupRef,
		Db			jdb)
	{
		m_this = lBackupRef;
		m_jdb = jdb;
	}

	long getRef()
	{
		return m_this;
	}
	
// PRIVATE METHODS
	
	private native void _release(
		long				lThis);
	
	private native long _getBackupTransId(
		long				lThis);
	
	private native long _getLastBackupTransId(
		long				lThis);
		
	private native long _backup(
		long				lThis,
		String			sBackupPath,
		String			sPassword,
		BackupClient	Client,
		BackupStatus	Status) throws XFlaimException;
	
	private native void _endBackup(
		long				lThis) throws XFlaimException;
	
	private long			m_this;
	private Db				m_jdb;
}

